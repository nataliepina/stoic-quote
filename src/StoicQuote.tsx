import { faTwitter } from "@fortawesome/free-brands-svg-icons";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import React, { useEffect, useRef, useState } from "react";
import Loader from "react-loader-spinner";
import "react-loader-spinner/dist/loader/css/react-spinner-loader.css";
import DisplayQuote from "./DisplayQuote";
import { copyText } from "./helpers/helpers";
import { Button, Main } from "./styledComponents";

interface QuoteData {
  quote: string;
  author: string;
}

const StoicQuote: React.FC = () => {
  const [quote, setQuote] = useState<QuoteData | null>(null);
  const [loading, setLoading] = useState(true);
  const [toolTip, setToolTip] = useState("Copy");
  const text = useRef("");

  useEffect(() => {
    const stoicQuote = require("stoic-quotes");

    const getQuote = async () => {
      try {
        setLoading(true);
        const res = await stoicQuote();
        const { quote, author } = res;

        setQuote({ quote, author });
        text.current = `"${quote}" -${author}`;
      } catch (e) {
        setQuote(null);
      } finally {
        setTimeout(() => setLoading(false), 500);
      }
    };
    if (!quote) getQuote();
  }, [quote]);

  const getNewQuote = () => {
    setQuote(null);
    setLoading(true);
  };

  const handleCopy = () => {
    copyText(text.current);
    setToolTip("Copied");
    setTimeout(() => setToolTip("Copy"), 1500);
  };

  return (
    <>
      <div className="Quote">
        <Main className="Quote-Main">
          {loading ? (
            <Loader type="ThreeDots" color="lightgray" height={80} width={80} />
          ) : (
            <DisplayQuote quote={quote} />
          )}
        </Main>
        <section className="Quote-Buttons">
          <Button onClick={getNewQuote}>New Quote</Button>
          <Button onClick={handleCopy}>{toolTip}</Button>
          <Button>
            <a
              id="tweet-quote"
              title="Tweet This Quote!"
              href={`https://twitter.com/intent/tweet?text=${text.current}`}
              target="_blank"
              rel="noopener noreferrer"
              className="Tweet-Button"
            >
              <FontAwesomeIcon icon={faTwitter} />
            </a>
          </Button>
        </section>
      </div>
    </>
  );
};

export default StoicQuote;
