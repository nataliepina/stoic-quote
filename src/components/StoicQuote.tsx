import { faTwitter } from '@fortawesome/free-brands-svg-icons';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import React, { useEffect, useRef, useState } from 'react';
import Loader from 'react-loader-spinner';
import 'react-loader-spinner/dist/loader/css/react-spinner-loader.css';
import styled from 'styled-components';
import { copyText } from '../helpers/helpers';
import DisplayQuote, { QuoteData } from './DisplayQuote';

export const QuoteContainer = styled.div`
  width: 75vw;
  margin: 1rem auto;
  display: flex;
  flex-direction: column;
  place-items: center center;
  font-family: 'Fraunces', serif;
  font-weight: 300;
  padding: 2.5rem 0;
  font-size: 1rem;
  font-weight: 600;
`;

export const ButtonWrapper = styled.div`
  margin: ${({ theme: { sizes } }) => sizes.xl} auto;
  width: 100%;
  display: flex;
  justify-content: center;
  gap: ${({ theme: { sizes } }) => sizes.md};
  font-family: 'Fraunces', serif;
`;

export const Button = styled.button`
  cursor: pointer;
  padding: 0.25rem 0.75rem;
  background-color: ${({ theme: { colors } }) => colors.dark};
  color: ${({ theme: { colors } }) => colors.light};
  font-size: ${({ theme: { fontSizes } }) => fontSizes.md};
  border: 2px solid ${({ theme: { colors } }) => colors.light};
  font-family: 'Cinzel', serif;

  &:hover {
    transition: all 0.8s;
    color: ${({ theme: { colors } }) => colors.primary};
    transform: scale(1, 1);
  }

  a {
    color: ${({ theme: { colors } }) => colors.light};

    &:hover {
      transition: all 0.6s;
      color: ${({ theme: { colors } }) => colors.primary};
      transform: scale(1, 1);
    }
  }
`;

const StoicQuote = (): React.ReactElement => {
  const [quote, setQuote] = useState<QuoteData | null>(null);
  const [loading, setLoading] = useState(true);
  const [toolTip, setToolTip] = useState('Copy');
  const text = useRef('');
  const newQuoteText = 'New Quote';

  useEffect(() => {
    // eslint-disable-next-line @typescript-eslint/no-var-requires
    const stoicQuote = require('stoic-quotes');

    const getQuote = async () => {
      try {
        setLoading(true);
        const res = await stoicQuote();
        const { quote, author } = res;

        setQuote({ quote, author });
        text.current = `"${quote}" -${author}`;
      } catch (e) {
        setQuote(null);
      } finally {
        setTimeout(() => setLoading(false), 500);
      }
    };
    if (!quote) getQuote();
  }, [quote]);

  const getNewQuote = () => {
    setQuote(null);
    setLoading(true);
  };

  const handleCopy = () => {
    copyText(text.current);
    setToolTip('Copied');
    setTimeout(() => setToolTip('Copy'), 1500);
  };

  return (
    <>
      <QuoteContainer>
        <section>
          {loading ? (
            <Loader type="ThreeDots" color="lightgray" height={80} width={80} />
          ) : (
            <DisplayQuote quote={quote} />
          )}
        </section>
        <ButtonWrapper>
          <Button onClick={getNewQuote}>{newQuoteText}</Button>
          <Button onClick={handleCopy}>{toolTip}</Button>
          <Button>
            <a
              id="tweet-quote"
              title="Tweet This Quote!"
              href={`https://twitter.com/intent/tweet?text=${text.current}`}
              target="_blank"
              rel="noopener noreferrer"
            >
              <FontAwesomeIcon icon={faTwitter} />
            </a>
          </Button>
        </ButtonWrapper>
      </QuoteContainer>
    </>
  );
};

export default StoicQuote;
