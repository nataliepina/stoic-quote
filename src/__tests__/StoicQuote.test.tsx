import React from 'react';
import StoicQuote from '../StoicQuote';
import { render } from '@testing-library/react';

describe('Stoic Quote tests', () => {
	it('should match snapshot', () => {
		const { asFragment } = render(<StoicQuote />);
		expect(asFragment()).toMatchSnapshot();
	});

	it('should render quote text', async () => {
		const { getByText } = render(<StoicQuote />);

		expect(getByText('Copy')).toBeInTheDocument();
    expect(getByText('New Quote')).toBeInTheDocument();
	});
});