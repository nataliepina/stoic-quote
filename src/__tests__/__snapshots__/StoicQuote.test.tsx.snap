// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Stoic Quote tests should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="Quote"
  >
    <main
      class="sc-bdfBwQ hHGPNm Quote-Main"
    >
      <div
        aria-busy="true"
        class=""
      >
        <svg
          aria-label="audio-loading"
          fill="lightgray"
          height="80"
          viewBox="0 0 120 30"
          width="80"
          xmlns="http://www.w3.org/2000/svg"
        >
          <circle
            cx="15"
            cy="15"
            r="15"
          >
            <animate
              attributeName="r"
              begin="0s"
              calcMode="linear"
              dur="0.8s"
              from="15"
              repeatCount="indefinite"
              to="15"
              values="15;9;15"
            />
            <animate
              attributeName="fillOpacity"
              begin="0s"
              calcMode="linear"
              dur="0.8s"
              from="1"
              repeatCount="indefinite"
              to="1"
              values="1;.5;1"
            />
          </circle>
          <circle
            attributeName="fillOpacity"
            cx="60"
            cy="15"
            from="1"
            r="9"
            to="0.3"
          >
            <animate
              attributeName="r"
              begin="0s"
              calcMode="linear"
              dur="0.8s"
              from="9"
              repeatCount="indefinite"
              to="9"
              values="9;15;9"
            />
            <animate
              attributeName="fillOpacity"
              begin="0s"
              calcMode="linear"
              dur="0.8s"
              from="0.5"
              repeatCount="indefinite"
              to="0.5"
              values=".5;1;.5"
            />
          </circle>
          <circle
            cx="105"
            cy="15"
            r="15"
          >
            <animate
              attributeName="r"
              begin="0s"
              calcMode="linear"
              dur="0.8s"
              from="15"
              repeatCount="indefinite"
              to="15"
              values="15;9;15"
            />
            <animate
              attributeName="fillOpacity"
              begin="0s"
              calcMode="linear"
              dur="0.8s"
              from="1"
              repeatCount="indefinite"
              to="1"
              values="1;.5;1"
            />
          </circle>
        </svg>
      </div>
    </main>
    <section
      class="Quote-Buttons"
    >
      <button
        class="sc-dlfnbm iaQAIo"
      >
        New Quote
      </button>
      <button
        class="sc-dlfnbm iaQAIo"
      >
        Copy
      </button>
      <button
        class="sc-dlfnbm iaQAIo"
      >
        <a
          class="Tweet-Button"
          href="https://twitter.com/intent/tweet?text="
          id="tweet-quote"
          rel="noopener noreferrer"
          target="_blank"
          title="Tweet This Quote!"
        >
          <svg
            aria-hidden="true"
            class="svg-inline--fa fa-twitter fa-w-16 "
            data-icon="twitter"
            data-prefix="fab"
            focusable="false"
            role="img"
            viewBox="0 0 512 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M459.37 151.716c.325 4.548.325 9.097.325 13.645 0 138.72-105.583 298.558-298.558 298.558-59.452 0-114.68-17.219-161.137-47.106 8.447.974 16.568 1.299 25.34 1.299 49.055 0 94.213-16.568 130.274-44.832-46.132-.975-84.792-31.188-98.112-72.772 6.498.974 12.995 1.624 19.818 1.624 9.421 0 18.843-1.3 27.614-3.573-48.081-9.747-84.143-51.98-84.143-102.985v-1.299c13.969 7.797 30.214 12.67 47.431 13.319-28.264-18.843-46.781-51.005-46.781-87.391 0-19.492 5.197-37.36 14.294-52.954 51.655 63.675 129.3 105.258 216.365 109.807-1.624-7.797-2.599-15.918-2.599-24.04 0-57.828 46.782-104.934 104.934-104.934 30.213 0 57.502 12.67 76.67 33.137 23.715-4.548 46.456-13.32 66.599-25.34-7.798 24.366-24.366 44.833-46.132 57.827 21.117-2.273 41.584-8.122 60.426-16.243-14.292 20.791-32.161 39.308-52.628 54.253z"
              fill="currentColor"
            />
          </svg>
        </a>
      </button>
    </section>
  </div>
</DocumentFragment>
`;
